### DISCUSSION TITLE
Detailed API

### DISCUSSION THREAD

**Discussion by jmcclure0921** [mKeRix/room-assistant#762] [Ideas]
Is there a detailed API of the MQTT messages or HTTP messages that room assistant uses to manage the cluster/devices?  I was looking to add on to the project by extending it to work on ESP32's.  Based on what I know of the project the difficulty level would be pretty low with my skill set and I think it would be a great feature.

**Comment by mKeRix** [comment]
It would be an awesome feature indeed! At the moment there is no API documentation for the internal messages, since they aren't really sent over a standardized protocol. I wanted to update this to use something like ZeroMQ sometime in the future, but haven't gotten around to it yet. If you already want to dig into it now you can take a look at the [cluster.service.ts](https://github.com/mKeRix/room-assistant/blob/main/src/cluster/cluster.service.ts) and the [democracy](https://www.npmjs.com/package/democracy) npm module. The packets are sent via UDP sockets, peer auto-discovery happens through mDNS.

Aside from that, you can already achieve some interoperability using `sendMqttRoom: true` ([docs](https://www.room-assistant.io/integrations/home-assistant.html#settings)), which will let Home Assistant match distance messages from multiple sources into a single sensor. You lose some of the more advanced functionality of room-assistant, but it should work ok as well.

**Comment by chedda61** [comment]
For another project of mine, I finally managed to get reliable rssi readings from my iphone using an esp32.
I reverse engineered the messages sent and received by room-assistant and hand crafted them for the esp-idf's hci API.
I can share the code if someone is willing to work on the rest to make it a room-assistant instance.
